[62992] Show single date work packages in the calendar

cbliard opened the pull request:
# Ticket

https://community.openproject.org/wp/62992

# What are you trying to accomplish?

In calendar module, show work packages even if they only have 1 date set.

# What approach did you choose and why?

For work package with end date only: use end date as start date if start date is absent.
For work package with start date only: they were already visible so nothing particular.

# Merge checklist

- [X] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

cbliard committed:
[62992] Show single date work packages in the calendar

https://community.openproject.org/wp/62992

HDinger commented on frontend/src/app/features/calendar/wp-calendar/wp-calendar.component.ts:
```diff
@@ -400,11 +400,11 @@ export class WorkPackagesCalendarComponent extends UntilDestroyedMixin implement
       return {
         title: workPackage.subject,
-        start: startDate,
+        start: startDate || endDate,
```
This feels really weird.. why should the start date be either start date or end date. Any future developer without the backgound of this ticket will fail to understand this. Please leave at least a comment.

cbliard replied:
Agreed, I have added a proper comment that it's the start date of the calendar event.

HDinger reviewed (commented):
Actually, the issue only happens when there is no start date set. I will update the ticket accordingly. 

With regards to the solution: It works as "specified" altough the AC are rather slim.. Currently there is no visual difference between a WP that has start and end at the same date vs a WP that only has one of them set. In the Gantt chart, we solve that with a gradient:

<img width="165" alt="Bildschirmfoto 2025-04-14 um 10 59 57" src="https://github.com/user-attachments/assets/70cced2a-6a7f-4c25-8af3-a8f3f53765ad" />

Imho this would make sense here as well. Currently in the Gantt chart, this is an inline style, but we could easily extract that into a class and apply it in both cases. Given however, that WP with only a start date have always been shown in the calendar and nobody complained about it before, is might not be such an issue after all 🤷

cbliard committed:
Comment why the end date is used for calendar work package events

cbliard commented:
> Imho this would make sense here as well. Currently in the Gantt chart, this is an inline style, but we could easily extract that into a class and apply it in both cases. **Given however, that WP with only a start date have always been shown in the calendar and nobody complained about it before, is might not be such an issue after all** 🤷

Yes, that was my conclusion too. That's probably good enough.
Maybe there could be another feature to add an indication that the start or finish date is not set. Gradient could be a solution.

cbliard commented:
@HDinger, I have added a comment as requested. Could you review it again please?

HDinger reviewed (approved):
I left a comment in the ticket for the gradient thing. I am fine with extracting that into a separate ticket if the product team agrees with the proposition.

cbliard commented:
The failing spec is `spec/features/work_packages/scheduling/automatic_scheduling_logic_spec.rb:340`.
It looks to be flickering because it passes locally.

cbliard committed:
Fix flickering spec

Failing spec is 'spec/features/work_packages/scheduling/automatic_scheduling_logic_spec.rb:340'
CI job run is https://github.com/opf/openproject/actions/runs/14442271942/job/40494880063?pr=18613

The spec was failing because the action menu was not visible (maybe it
was hidden by a relations tab refresh).

The fix is to retry the action until it works.
